src/finufft_utils.cpp

```diff
@@ -57,6 +70,102 @@ double CNTime::elapsedsec() const
   return nowsec - initial;
 }
 
+// Returns the number of physical CPU cores (excluding hyper-threaded logical cores)
+int getPhysicalCoreCount() {
+  int physicalCoreCount = 0;
+
+#if defined(_WIN32)
+  // Determine the size of the buffer.
+  DWORD bufferSize = 0;
+  if (GetLogicalProcessorInformation(nullptr, &bufferSize) == FALSE &&
+      GetLastError() != ERROR_INSUFFICIENT_BUFFER) {
+    // Leave physicalCoreCount as 0 if the call fails.
+    return physicalCoreCount;
+  }
+
+  // Calculate the number of entries and allocate a vector.
+  size_t entryCount = bufferSize / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION);
+  std::vector<SYSTEM_LOGICAL_PROCESSOR_INFORMATION> procInfo(entryCount);
+  if (GetLogicalProcessorInformation(procInfo.data(), &bufferSize) != FALSE) {
+    for (const auto &info : procInfo) {
+      if (info.Relationship == RelationProcessorCore) ++physicalCoreCount;
+    }
+  }
+
+#elif defined(__APPLE__)
+  // Retrieve the number of physical cores using sysctl.
+  int cores   = 0;
+  size_t size = sizeof(cores);
+  if (sysctlbyname("hw.physicalcpu", &cores, &size, nullptr, 0) == 0)
+    physicalCoreCount = static_cast<unsigned int>(cores);
+
+#elif defined(__linux__)
```

Comment:
Just a small word of caution: this will always report the full umber of physical cores on the system, independent of possible restrictions on the executable itself. On computing cener nodes, the system may limit a job to a part of the available hardware (e.g. via `taskset`); if this happens, the code here will overestimate the number of usable cores.
On systems with power and efficiency cores the code will probably (this is just speculation, I haven't tested it) return the sum of both core counts, which may also be larger than the amount of cores that should actually be used.

I'm not saying that the code should be changed (I wouldn't know how...), but this could be worth remembering in case there are reports about weird behaviour in the future.

Reply:
I was going to suggest looking at `hwloc` for inspiration, and it looks like you've found a very similar solution that they did. For posterity:
https://github.com/open-mpi/hwloc/blob/1779eae26f2f55a510e9d73cb648b68e71d9e6d4/hwloc/topology-linux.c#L1036

Reply:
I was going to suggest `libpthreads` ... it seems this functionality is available everywhere these days :-)

Reply:
Where is the functionality in pthread?  I didn't go the route of hwloc because I read somewhere that is an issue on maybe freebsd? This seemed more portable to me

Reply:
In `ducc` I'm using
```C++
size_t available_hardware_threads()
  {
  static const size_t available_hardware_threads_ = []()
    {
#if __has_include(<pthread.h>) && defined(__linux__) && defined(_GNU_SOURCE)
    cpu_set_t cpuset;
    CPU_ZERO(&cpuset);
    pthread_getaffinity_np(pthread_self(), sizeof(cpuset), &cpuset);
    size_t res=0;
    for (size_t i=0; i<CPU_SETSIZE; ++i)
      if (CPU_ISSET(i, &cpuset)) ++res;
#else
    size_t res = std::max<size_t>(1, std::thread::hardware_concurrency());
#endif
    return res;
    }();
  return available_hardware_threads_;
  }
```

which is very close to your `getAllowedCoreCount` function, but it gets the affinity masks from `libpthreads` instad of the GNU `libc`. It *might* be more portable, but I'm not sure.